File: src/main/java/io/github/ajoz/workshop/fp/part_1/solutions/exercise_6/Exercise6.java
package io.github.ajoz.workshop.fp.part_1.solutions.exercise_6;

final class Customer {
    private final String name;

    Customer(final String name) {
        this.name = name;
    }
}

final class Order {
    final String title;
    private final Long date;

    Order(String title, Long date) {
        this.title = title;
        this.date = date;
    }
}

final class Database {
    Order findOrder(final Customer customer) {
        return new Order("JUG Łódź -- visit our FB, twitter and meetup!", 42L);
    }
}

@FunctionalInterface
interface Function1<A, B> {
    B apply(A a);

    default <C> Function1<A, C> andThen(final Function1<B, C> after) {
        return a -> after.apply(this.apply(a));
    }
}

@FunctionalInterface
interface Function2<A, B, C> {
    C apply(A a, B b);
}

@FunctionalInterface
interface Supplier<A> {
    A get();
}

class Exercise6 {
    private static final Function2<Customer, Database, Order> getOrderForCustomer =
            (customer, database) -> database.findOrder(customer);

    private static final Supplier<Database> getProductionDatabase = Database::new;

    private static final Function1<Order, String> getOrderTitle = order -> order.title;

    private static final Function1<String, Long> getHash = value -> 42L;

    public static Function1<Customer, Long> getCustomerToHash() {
        // flip the arguments of the original
        // change it to curried form
        // partially apply the first argument
        // change the Order to a title
        // change the title to a secret hash
        return applyFirst(curry(flip(getOrderForCustomer)), getProductionDatabase)
                .andThen(getOrderTitle)
                .andThen(getHash);
    }

    private static <A, B, C> Function1<A, Function1<B, C>> curry(final Function2<A, B, C> function2) {
        return (A a) -> (B b) -> function2.apply(a, b);
    }

    private static <A, B, C> Function2<B, A, C> flip(final Function2<A, B, C> function2) {
        return (B b, A a) -> function2.apply(a, b);
    }

    private static <A, B, C> Function1<B, C> applyFirst(final Function1<A, Function1<B, C>> function,
                                                        final Supplier<A> supplier) {
        return (B b) -> function.apply(supplier.get()).apply(b);
    }
}

File: src/main/kotlin/io/github/ajoz/workshop/fp/kotlin/part_1/solutions/exercise_6/Exercise6.kt
@file:Suppress("PackageName", "unused")

package io.github.ajoz.workshop.fp.kotlin.part_1.solutions.exercise_6

import io.github.ajoz.workshop.fp.kotlin.tools.andThen
import io.github.ajoz.workshop.fp.kotlin.tools.applySecond

// other fields like: surname, address, etc.
data class Customer(val name: String)

// other fields like: amount, currency, tax, etc.
data class Order(val title: Title, val date: Timestamp)

data class Title(val title: String)

data class Timestamp(val unixTimestamp: Long)

data class Hash(val value: Long)

internal class Database {
    // for now let's assume that the DB will always return an Order for a
    // given Customer
    fun findOrder(customer: Customer): Order {
        return Order(
                Title(String.format("FP Workshop - %s", customer.name)),
                Timestamp(42L)
        )
    }
}

object Exercise6 {
    // Please do not change this function!
    private val getOrderForCustomer: (Customer, Database) -> Order =
            { customer, database ->
                database.findOrder(customer)
            }

    // Please do not change this function!
    private val getProductionDatabase =
            { Database() }

    // Please do not change this function!
    private val getOrderTitle: (Order) -> Title =
            { it.title }

    // Please do not change this function!
    private val getTitleHash: (Title) -> Hash =
            {
                Hash(it.title.length.toLong())
            }

    /*
    // this version is using flipping, currying, application of the first argument
    // and composition
    fun customerToHash(): (Customer) -> Hash =
            getOrderForCustomer
                    .flipped()
                    .curried()
                    .apF(getProductionDatabase)
                    .andThen(getOrderTitle)
                    .andThen(getTitleHash)
    */

    /*
    // this version is using currying, application of the second argument
    // and composition
    fun customerToHash(): (Customer) -> Hash =
            getOrderForCustomer
                    .curried()
                    .apS(getProductionDatabase)
                    .andThen(getOrderTitle)
                    .andThen(getTitleHash)
    */

    /*
    // this version is using flipping, application of the first argument
    // and composition
    fun customerToHash(): (Customer) -> Hash =
            getOrderForCustomer
                    .flipped()
                    .applyFirst(getProductionDatabase)
                    .andThen(getOrderTitle)
                    .andThen(getTitleHash)
    */

    /*
    // this is probably the easiest to grasp and read in Kotlin version
    // for anyone that is not used to functional programming
    fun customerToHash(): (Customer) -> Hash =
            getOrderForCustomer                          //(Customer, Database) -> Order
                    .applySecond(getProductionDatabase)  // Customer -> Order
                    .andThen(getOrderTitle)              // Customer -> Title
                    .andThen(getTitleHash)               // Customer -> Hash
    */

    // with infix notation without the dot and parenthesis noise
    // unfortunately Kotlin doesn't make it easier to split it into multiple lines
    fun customerToHash(): (Customer) -> Hash =
            getOrderForCustomer applySecond getProductionDatabase andThen getOrderTitle andThen getTitleHash
}
